test-utils/src/main/kotlin/com/google/devtools/ksp/testutils/AbstractKSPTest.kt

```diff
@@ -0,0 +1,206 @@
+            +LanguageSettingsDirectives.ALLOW_KOTLIN_PACKAGE
+        }
+
+        configureTest(this)
+
+        startingArtifactFactory = { ResultingArtifact.Source() }
+        this.testInfo = this@AbstractKSPTest.testInfo
+    }
+
+    fun TestModule.loadKtFiles(project: Project): List<KtFile> {
+        return files.filter { it.isKtFile }.map {
+            KtTestUtil.createFile(it.name, it.originalContent, project)
+        }
+    }
+
+    fun TestModule.writeJavaFiles(): List<File> {
+        javaDir.mkdirs()
+        val files = javaFiles.map { it to File(javaDir, it.relativePath) }
+        files.forEach { (testFile, file) ->
+            file.parentFile.mkdirs()
+            file.writeText(testFile.originalContent)
+        }
+        return files.map { it.second }
+    }
+
+    // No, this is far from complete. It only works for our test cases.
+    //
+    // No, neither CompiledLibraryProvider nor LibraryEnvironmentConfigurator can be used. They rely on
+    // dist/kotlinc/lib/*
+    //
+    // No, sourceFileProvider doesn't group files by module unfortunately. Let's do it by ourselves.
+    fun compileModule(module: TestModule, testServices: TestServices) {
+        val javaFiles = module.writeJavaFiles()
+        val compilerConfiguration = testServices.compilerConfigurationProvider.getCompilerConfiguration(module)
+        val dependencies = module.allDependencies.map { outDirForModule(it.moduleName) }
+        compilerConfiguration.addJvmClasspathRoots(dependencies)
+        compilerConfiguration.addJavaSourceRoot(module.javaDir)
+
+        // TODO: other platforms
+        val kotlinCoreEnvironment = KotlinCoreEnvironment.createForTests(
+            disposable,
+            compilerConfiguration,
+            EnvironmentConfigFiles.JVM_CONFIG_FILES
+        )
+        val ktFiles = module.loadKtFiles(kotlinCoreEnvironment.project)
+        GenerationUtils.compileFilesTo(ktFiles, kotlinCoreEnvironment, module.outDir)
+
+        if (module.javaFiles.isEmpty())
+            return
+
+        val classpath = (dependencies + KtTestUtil.getAnnotationsJar() + module.outDir)
+            .joinToString(File.pathSeparator) { it.absolutePath }
+        val options = listOf(
+            "-classpath", classpath,
+            "-d", module.outDir.path
+        )
+        compileJavaFiles(javaFiles, options, assertions = JUnit5Assertions)
+    }
+
+    fun runTest(@TestDataFile path: String) {
+        val testConfiguration = testConfiguration(path, configure)
+        Disposer.register(disposable, testConfiguration.rootDisposable)
+        val testServices = testConfiguration.testServices
+        val moduleStructure = testConfiguration.moduleStructureExtractor.splitTestDataByModules(
+            path,
+            testConfiguration.directives,
+        )
+        val mainModule = moduleStructure.modules.last()
+        val libModules = moduleStructure.modules.filterNot { it == mainModule }
```

Comment:
nit: if the last module is always the main module, maybe just `dropLast(1)` ?

Reply:
Done